### PR TITLE
[syncd] Add pre match logic for pfc storm

### PR DESCRIPTION
**Pull request by kcudnik**
Special case where priority is the same for all acl

### CODE REVIEW

File: syncd/ComparisonLogic.cpp [hunk 1]
```diff
@@ -2836,6 +2838,21 @@ void ComparisonLogic::cretePreMatchForAclEntries(
                 continue;
             }
 
+            if (cPrio->getStrAttrValue() == PFC_ACL_RULE_PRIORITY)
+            {
+                // this is special case, all pfc rules to prevent packet storm
+                // will have same priority, we will need to check SAI_ACL_ENTRY_ATTR_FIELD_TC
+
+                auto cTc = cAclEntry->getSaiAttr(SAI_ACL_ENTRY_ATTR_FIELD_TC);
```
Comment (vaibhavhd) [inline syncd/ComparisonLogic.cpp:2846]: Will this get call fail if the TC field is not present in the ACL?

File: syncd/ComparisonLogic.cpp [hunk 2]
```diff
@@ -2806,6 +2806,8 @@ void ComparisonLogic::cretePreMatchForLagMembers(
     }
 }
 
+#define PFC_ACL_RULE_PRIORITY "999"
```
Comment (vaibhavhd) [inline syncd/ComparisonLogic.cpp:2809]: @neethajohn can you confirm if hardcoding 999 as a PFC ACL priority will work across branches?

File: syncd/ComparisonLogic.cpp [hunk 3]
```diff
@@ -2836,6 +2838,21 @@ void ComparisonLogic::cretePreMatchForAclEntries(
                 continue;
             }
 
+            if (cPrio->getStrAttrValue() == PFC_ACL_RULE_PRIORITY)
```
Comment (vaibhavhd) [inline syncd/ComparisonLogic.cpp:2841]: Do we know that only PFC ACLs have the priority of `PFC_ACL_RULE_PRIORITY`? If not, then there is a chance that this block will get executed for other ACLs, and in that case the behavior is indeterministic.

File: syncd/ComparisonLogic.cpp [hunk 4]
```diff
@@ -2836,6 +2838,21 @@ void ComparisonLogic::cretePreMatchForAclEntries(
...
+            {
+                // this is special case, all pfc rules to prevent packet storm
+                // will have same priority, we will need to check SAI_ACL_ENTRY_ATTR_FIELD_TC
+
+                auto cTc = cAclEntry->getSaiAttr(SAI_ACL_ENTRY_ATTR_FIELD_TC);
+                auto tTc = tAclEntry->getSaiAttr(SAI_ACL_ENTRY_ATTR_FIELD_TC);
+
+                if (!cTc || !cTc || cTc->getStrAttrValue() != tTc->getStrAttrValue())
```
Comment (vaibhavhd) [inline syncd/ComparisonLogic.cpp:2849]: I think you mean to check for `!cTc || !tTc` rather`!cTc || !cTc`?

File: syncd/ComparisonLogic.cpp [hunk 5]
```diff
@@ -2836,6 +2838,21 @@ void ComparisonLogic::cretePreMatchForAclEntries(
...
+                {
+                    // TC attribute is different, not looking for this one
+                    continue;
+                }
+            }
+
             // at this point current and temporary acl entry share the same
             // priority, then we can assume that those objects are the same
```
Comment (vaibhavhd) [inline syncd/ComparisonLogic.cpp:2857]: I think this comment needs to be updated? Now not just priority, but also TC field may be same.

File: syncd/ComparisonLogic.cpp (hunk 1 above)
Comment (kcudnik) [reply syncd/ComparisonLogic.cpp:2846]: yes, let me fix that

File: syncd/ComparisonLogic.cpp (hunk 2 above)
Comment (kcudnik) [reply syncd/ComparisonLogic.cpp:2809]: no clue, this is hardcoded as well in OA

File: syncd/ComparisonLogic.cpp (hunk 3 above)
Comment (kcudnik) [reply syncd/ComparisonLogic.cpp:2841]: this is hardcoded in OA https://github.com/sonic-net/sonic-swss/blob/f31ccd09b8eafa2d9292e7446fc5fe8d4ec0ad06/orchagent/pfcactionhandler.cpp#L448

File: syncd/ComparisonLogic.cpp (hunk 4 above)
Comment (kcudnik) [reply syncd/ComparisonLogic.cpp:2849]: yes

File: syncd/ComparisonLogic.cpp (hunk 5 above)
Comment (kcudnik) [reply syncd/ComparisonLogic.cpp:2857]: updated

**Comment by kcudnik** [comment]
Special case where priority is the same for all acl

**Comment by StormLiangMS** [comment]
@kcudnik cherry pick to 202305 conflict, could you help to raise separate PR for 202305?

**Comment by kcudnik** [comment]
sure

**Comment by kcudnik** [comment]
here it is: https://github.com/sonic-net/sonic-sairedis/pull/1318 please review

**Pull request merged**